File: src/nostr/filter.rs
use serde::{Deserialize, Serialize};
use crate::nostr::Event;

#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq, Eq, Hash)]
pub struct Filter {
    pub ids: Option<Vec<String>>,
    pub authors: Option<Vec<String>>,
    pub kinds: Option<Vec<u64>>,
    #[serde(rename = "#p")]
    pub p_tags: Option<Vec<String>>,
    #[serde(rename = "#e")]
    pub e_tags: Option<Vec<String>>,
    pub since: Option<u64>,
    pub until: Option<u64>,
    pub limit: Option<u64>,
}

impl Filter {
    pub fn matches(&self, event: &Event) -> bool {
        if let Some(ids) = &self.ids {
            if !ids.contains(&event.id) {
                return false;
            }
        }
        if let Some(authors) = &self.authors {
            if !authors.contains(&event.pubkey) {
                return false;
            }
        }
        if let Some(kinds) = &self.kinds {
            if !kinds.contains(&event.kind) {
                return false;
            }
        }
        if let Some(p_tags) = &self.p_tags {
            let has_match = event.tags.iter().any(|t| {
                t.pubkey().is_some_and(|pk| p_tags.iter().any(|s| s == pk))
            });
            if !has_match {
                return false;
            }
        }
        if let Some(e_tags) = &self.e_tags {
            let has_match = event.tags.iter().any(|t| {
                t.event_id().is_some_and(|eid| e_tags.iter().any(|s| s == eid))
            });
            if !has_match {
                return false;
            }
        }
        if let Some(since) = self.since {
            if event.created_at < since {
                return false;
            }
        }
        if let Some(until) = self.until {
            if event.created_at > until {
                return false;
            }
        }
        true
    }

    pub fn is_valid(&self) -> bool {
        let has_specific_narrowing = self.p_tags.is_some() || self.e_tags.is_some() || self.ids.is_some();

        if has_specific_narrowing {
            if let Some(kinds) = &self.kinds {
                if kinds.iter().any(|k| !matches!(k, 5 | 13194 | 23194..=23197)) {
                    return false;
                }
            }
        } else {
            let kinds = match &self.kinds {
                Some(k) if !k.is_empty() => k,
                _ => return false,
            };

            if kinds.iter().any(|k| !matches!(k, 5 | 13194 | 23194..=23197)) {
                return false;
            }

            if self.ids.is_none() && self.authors.is_none() && self.p_tags.is_none() && self.e_tags.is_none() {
                return false;
            }
        }

        true
    }

    pub fn pubkeys(&self) -> Vec<String> {
        let mut keys = Vec::new();
        if let Some(authors) = &self.authors {
            keys.extend(authors.clone());
        }
        if let Some(p_tags) = &self.p_tags {
            keys.extend(p_tags.clone());
        }
        keys
    }
}

#[cfg(test)]
#[path = "filter_test.rs"]
mod filter_test;

File: src/nostr/filter.ts
import { Event } from './event.ts';

export interface Filter {
  ids?: string[];
  authors?: string[];
  kinds?: number[];
  pTags?: string[];
  eTags?: string[];
  since?: number;
  until?: number;
  limit?: number;
}

const NWC_KINDS = new Set([5, 13194, 23194, 23195, 23196, 23197]);

function isNwcKind(k: number): boolean {
  return NWC_KINDS.has(k);
}

export function filterFromJSON(json: Record<string, unknown>): Filter {
  const f: Filter = {};
  if (Array.isArray(json.ids)) f.ids = json.ids as string[];
  if (Array.isArray(json.authors)) f.authors = json.authors as string[];
  if (Array.isArray(json.kinds)) f.kinds = json.kinds as number[];
  const pRaw = json['#p'];
  if (Array.isArray(pRaw)) f.pTags = pRaw as string[];
  const eRaw = json['#e'];
  if (Array.isArray(eRaw)) f.eTags = eRaw as string[];
  if (typeof json.since === 'number') f.since = json.since as number;
  if (typeof json.until === 'number') f.until = json.until as number;
  if (typeof json.limit === 'number' || typeof json.limit === 'string') {
    f.limit = Number(json.limit);
  }
  return f;
}

export function filterToJSON(f: Filter): Record<string, unknown> {
  const json: Record<string, unknown> = {};
  if (f.ids) json.ids = f.ids;
  if (f.authors) json.authors = f.authors;
  if (f.kinds) json.kinds = f.kinds;
  if (f.pTags) json['#p'] = f.pTags;
  if (f.eTags) json['#e'] = f.eTags;
  if (f.since !== undefined) json.since = f.since;
  if (f.until !== undefined) json.until = f.until;
  if (f.limit !== undefined) json.limit = f.limit;
  return json;
}

export function filterMatches(filter: Filter, event: Event): boolean {
  if (filter.ids && !filter.ids.includes(event.id)) {
    return false;
  }
  if (filter.authors && !filter.authors.includes(event.pubkey)) {
    return false;
  }
  if (filter.kinds && !filter.kinds.includes(event.kind)) {
    return false;
  }
  if (filter.pTags) {
    const hasMatch = event.tags.some(t => {
      const pk = t.pubkey();
      return pk !== null && filter.pTags!.includes(pk);
    });
    if (!hasMatch) return false;
  }
  if (filter.eTags) {
    const hasMatch = event.tags.some(t => {
      const eid = t.eventId();
      return eid !== null && filter.eTags!.includes(eid);
    });
    if (!hasMatch) return false;
  }
  if (filter.since !== undefined && event.createdAt < filter.since) {
    return false;
  }
  if (filter.until !== undefined && event.createdAt > filter.until) {
    return false;
  }
  return true;
}

export function filterIsValid(filter: Filter): boolean {
  const hasSpecificNarrowing =
    filter.pTags !== undefined ||
    filter.eTags !== undefined ||
    filter.ids !== undefined;

  if (hasSpecificNarrowing) {
    if (filter.kinds) {
      if (filter.kinds.some(k => !isNwcKind(k))) {
        return false;
      }
    }
  } else {
    const kinds = filter.kinds;
    if (!kinds || kinds.length === 0) return false;
    if (kinds.some(k => !isNwcKind(k))) return false;
    if (
      filter.ids === undefined &&
      filter.authors === undefined &&
      filter.pTags === undefined &&
      filter.eTags === undefined
    ) {
      return false;
    }
  }
  return true;
}

export function filterPubkeys(filter: Filter): string[] {
  const keys: string[] = [];
  if (filter.authors) keys.push(...filter.authors);
  if (filter.pTags) keys.push(...filter.pTags);
  return keys;
}
